website/content/en/docs/kubebuilder-migration/project_migration_guide.md

```diff
@@ -0,0 +1,275 @@
+
+```go
+type MemcachedSpec struct {
+	// Specify size, the number of pods
+	Size int32 `json:"size"`
+}
+
+type MemcachedStatus struct {
+	// Nodes are the names of the memcached pods
+	Nodes []string `json:"nodes"`
+}
+```
+**Note**: 
+If there are any any libraries or pkgs present in `pkg/apis/cache/v1alpha1`, copy them over to `api/v1alpha1`.
+
+### CRD markers
+
+Include any of the CRD generation or validation markers defined for your API type and fields in `api/<version>/<kind>_types.go`.
+
+For `memcached-operator` the following markers present in `pkg/apis/cache/v1aplha1/memcached_types.go` are copied to `api/v1alpha1/memcached_types.go`.
+
+Example:
+```Go
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=memcacheds,scope=Namespaced
+```
+
+### Migrate controller
+
+Copy over the `Reconcile()` code and the controller logic from your existing controller at `pkg/controller/<kind>/<kind>_controller.go` to the new controller's reconcile function in Kubebuilder project at `controllers/<kind>_controller.go`.
+
+In our example [`pkg/controller/memcached/memcached_controller.go`][memcached_controller] is copied to `controllers/memcached_controller.go`.
+
+Particularly, note the following changes:
+
+1. The naming convention of the reconciler struct may differ. For example, in case of memcached operator the reconciler struct in SDK project is mentioned as [`ReconcileMemcached`][sdk_reconcile_struct] whereas in Kubebuilder project it would be `MemcachedReconciler`.
+
+```go
+type MemcachedReconciler struct {
+	Log    logr.Logger
+	Client client.Client
+	Scheme *runtime.Scheme
+}
+```
+
+2. Update the import aliases in `Reconcile` logic and copy over the remaining related functions from `pkg/controller/<kind>/<kind>_controller.go` to `api/<version>/<kind>_controller.go`.
+
+Refer to [`memcached_controller.go`][kb_memcached_controller] for the example implementation of `reconcile` logic which would be present in `api/v1alpha1/memcached_controller.go` of Memcached operator.
+
+## Generate CRDs
+
+Run [`make manifests`][generate_crd] to generate CRD manifests. They would be generated inside the `config/` folder.
+
+```sh
+$ make manifests
+go: creating new go.mod: module tmp
+go: finding sigs.k8s.io v0.2.5
+go: finding sigs.k8s.io/controller-tools/cmd v0.2.5
+go: finding sigs.k8s.io/controller-tools/cmd/controller-gen v0.2.5
+/Users/admin/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+```
+
+## Operator Manifests
+
+### Operator deployment manifests
+
+The operator deployment manifest [`operator.yaml`][deployment_yaml] present inside `deploy/` folder in SDK project can be copied to `config/manager/manager.yaml`
+
+**Note**: 
+The kustomize file requires the operator deployment manifest to have the field `namespace` which is missing in the current operator.yaml of sdk project.
```

Comment:
This probably gets into a little more beyond the migration guide so maybe you can add a TODO here but I think when users copy over their deployment they'll want to make sure they keep their Deployment selector labels the same as the default one `control-plane: controller-manager` as in the metrics service.
https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/manager/manager.yaml#L18
https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/rbac/auth_proxy_service.yaml#L14
https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/prometheus/monitor.yaml#L16

Otherwise the service and servicemonitor won't be able to scrape the operator metrics I think.
https://book.kubebuilder.io/reference/metrics.html

So this is probably a follow up we need to do, to double check that we need to preserve the default `control-plane: controller-manager` labels so that the metrics are correctly exposed.

Reply:
Yes, if the labels for pods and deployment is modified, then [this](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/rbac/auth_proxy_service.yaml#L14) should be modified. And if the label for [service](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/rbac/auth_proxy_service.yaml#L5) changes, it should be modified in [service monitor](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/prometheus/monitor.yaml#L16) too. Will add it as TODO as of now.